internal/kubeobjects/redaction.go

```diff
@@ -0,0 +1,51 @@
+package kubeobjects
+
+import (
+	"strings"
+)
+
+// `expr` is a dot-separated address for nested values, in the same format as
+// arguments to `kubectl explain`.
+// For example, the expr "spec.containers.env" will cause Kubernetes Pod
+// container environment variables to be redacted. "containers" is an array, and
+// implicitly each element of this array is redacted.
+func Redact(obj interface{}, expr string) interface{} {
+	exprParts := strings.Split(expr, ".")
+	return redact(obj, exprParts)
+}
+
+func redact(obj interface{}, exprParts []string) interface{} {
```

Comment:
I am not too attached to this hand-rolled implementation. I think the tests do give us some confidence, but the real reason I wrote it is the failure to find 3rd party libraries above. After I'd written it, @srlk pointed me to https://github.com/itchyny/gojq. It does look reasonably battle-tested, and I'm not against replacing this implementation (and redaction expression syntax) with it.

As well as the option in the other thread, to use jsonpatch but require users to be precise about exact fields to clip out, without being able to express concepts like "all pod spec env vars", I think gojq is option 3 (option 1 being "keep the code as written").

Reply:
I think there's a *much simpler* way to do this - `unstructured.Unstructured` already supports this:
https://pkg.go.dev/k8s.io/apimachinery/pkg/apis/meta/v1/unstructured#RemoveNestedField

(Haven't tested it, but as far as I can tell, you could just do:
```go
unstructured.RemoveNestedField(obj.Object, strings.Split(expr, ".")...)
```

Reply:
Also - no offense - but I much prefer the [implementation](https://github.com/kubernetes/apimachinery/blob/v0.27.1/pkg/apis/meta/v1/unstructured/helpers.go#L265) of `RemoveNestedField`:
```go
func RemoveNestedField(obj map[string]interface{}, fields ...string) {
	m := obj
	for _, field := range fields[:len(fields)-1] {
		if x, ok := m[field].(map[string]interface{}); ok {
			m = x
		} else {
			return
		}
	}
	delete(m, fields[len(fields)-1])
}
```

Reply:
Oh wow, I really wish I had seen that! Does it support removing object fields nested inside array elements?

Reply:
I'm not seeing how those functions help, not quickly at least... if this is something already in your head, could you push a branch on top of this one showing me what you mean? Take a look at the test suite first, if we agree on that, then keeping it passing should prove if it works.

AFAIK the 2 major differences between my impl and unstructured.RemoveNestedField are nested array operations (what we're discussing), and recursion vs loops (which I don't have a strong opinion on - I think I just tend to reach for recursion for things that traverse structures). Holler if there is an aspect I'm missing.

I think the bigger picture is actually in the other thread I started - should we use this `kubectl explain` syntax and allow ops on "all" array elements at all? If that one tips to "no", this thread becomes moot. But I really am not sure about that one yet.

Reply:
> should we use this kubectl explain syntax and allow ops on "all" array elements at all?

I like this for its simplicity

I doubt anyone wants to do something like `spec.containers[1].blabla`

Reply:
Yeah sorry for the noise - I just had a look at your full implementation and it's nice! :) 

Regarding all array elements: yes I'd say so. Explicit indexes would be a bit weird, as Serol noted.

I don't see why we shouldn't keep this 👍

Reply:
@srlk I haven't fully validated this but with jmespath specifically it looks like we might be able to index by field, e.g. `spec.containers[name=my-app].env`. But again, not "all", which might be our modal use case (a bit of a guess)